src/common/tests_dmg_helpers.c

```diff
@@ -141,88 +142,156 @@ log_stderr_pipe(int fd)
+{
+	char	*tmp_cmd;
+	char	*ch;
+	char	*saveptr = NULL;
+	char	**tmp = NULL;
+	char	**result = NULL;
+	int	result_len = 0;
+	int	rc = 0;
+
+	D_STRNDUP(tmp_cmd, cmd, strnlen(cmd, 1024));
+	if (tmp_cmd == NULL)
+		return -DER_NOMEM;
+
+	ch = strtok_r(tmp_cmd, " ", &saveptr);
+	while (ch != NULL) {
+		result = cmd_push_arg(result, &result_len, ch);
+		if (result == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+		ch = strtok_r(NULL, " ", &saveptr);
+	}
+
+	D_ASSERT(result_len > 1);
+
+	/* last item of the array must be NULL for use with exec */
+	D_REALLOC_ARRAY(tmp, result, result_len, result_len + 1);
+	if (tmp == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	result = tmp;
+	result[result_len] = NULL;
+	result_len++;
+
+	*tokens = result;
+	*tokens_len = result_len;
+
+out:
+	D_FREE(tmp_cmd);
+	return rc;
+}
+
 static int
 run_cmd(const char *command, int *outputfd)
 {
-	int rc = 0;
-	int child_rc = 0;
-	int child_pid;
-	int stdoutfd[2];
-	int stderrfd[2];
+	int	rc = 0;
+	int	child_rc = 0;
+	int	child_pid;
+	int	stdoutfd[2];
+	int	stderrfd[2];
+	char	**cmd_tokens = NULL;
+	int	tokens_len = 0;
+	char	*file = NULL;
 
 	D_DEBUG(DB_TEST, "dmg cmd: %s\n", command);
 
 	/* Create pipes */
 	if (pipe(stdoutfd) == -1) {
 		rc = daos_errno2der(errno);
 		D_ERROR("failed to create stdout pipe: %s\n", strerror(errno));
-		return rc;
+		D_GOTO(err, rc);
 	}
 
 	if (pipe(stderrfd) == -1) {
 		rc = daos_errno2der(errno);
 		D_ERROR("failed to create stderr pipe: %s\n", strerror(errno));
-		close(stdoutfd[0]);
-		close(stdoutfd[1]);
-		return rc;
+		D_GOTO(err_stdout, rc);
 	}
 
+	rc = cmd2tokens(command, &cmd_tokens, &tokens_len);
+	if (rc != 0) {
+		D_ERROR("failed to split command into tokens: "DF_RC"\n", DP_RC(rc));
+		D_GOTO(err_stderr, rc);
+	}
+
+	D_ASPRINTF(file, cmd_tokens[0]);
+	if (file == NULL)
+		D_GOTO(err_tokens, rc = -DER_NOMEM);
+
+	D_DEBUG(DB_TEST, "forking to run dmg command\n");
+
 	child_pid = fork();
 	if (child_pid == -1) {
 		rc = daos_errno2der(errno);
 		D_ERROR("failed to fork: %s\n", strerror(errno));
-		return rc;
+		D_GOTO(err_tokens, rc);
 	}
-
-	if (child_pid == 0) {
+	else if (child_pid == 0) {
 		/* child doesn't need the read end of the pipes */
 		close(stdoutfd[0]);
 		close(stderrfd[0]);
 
-		D_DEBUG(DB_TEST, "running dmg command\n");
-
-		if (dup2(stdoutfd[1], STDOUT_FILENO) == -1) {
-			D_ERROR("failed to dup stdout pipe: %s\n", strerror(errno));
-			_exit(daos_errno2der(errno));
-		}
+		if (dup2(stderrfd[1], STDERR_FILENO) == -1)
+			_exit(errno);
 
-		if (dup2(stderrfd[1], STDERR_FILENO) == -1) {
-			D_ERROR("failed to dup stderr pipe: %s\n", strerror(errno));
-			_exit(daos_errno2der(errno));
-		}
+		if (dup2(stdoutfd[1], STDOUT_FILENO) == -1)
+			_exit(errno);
 
 		close(stdoutfd[1]);
 		close(stderrfd[1]);
 
-		if (system(command) == -1) {
-			D_ERROR("failed to invoke '%s', errno=%d (%s)\n", command, errno,
-				strerror(errno));
-			_exit(daos_errno2der(errno));
+		rc = execvp(file, cmd_tokens);
+		if (rc == -1) {
```

Comment:
That is true. I went back and forth on what to do here. Maybe just assert rc == -1 and return the errno, to avoid thinking about things that can't happen.

Reply:
Fixed.